[CSGen] Allow `_` pattern to be a hole

xedin opened the pull request:
In cases like `case .test(_)` it might not be possible to
establish a type of `_` and hole cannot be propagated to
it from context if condition of switch is incorrect, so
`_` just like a named pattern should be allowed to become
a hole.

Resolves: rdar://96997534

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintLocator] Generalize pattern decl locator to support "named" and "any"

xedin committed:
[CSFix] Generalize a fix for unresolved pattern decl

The fix should support both named (i.e. `test(a)` and "any" patterns
i.e. `test(_)`.

xedin committed:
[CSGen] Allow `_` pattern to be a hole

In cases like `case .test(_)` it might not be possible to
establish a type of `_` and hole cannot be propagated to
it from context if condition of switch is incorrect, so
`_` just like a named pattern should be allowed to become
a hole.

Resolves: rdar://96997534

xedin commented:
@swift-ci please test

ahoppen commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -5275,6 +5275,12 @@ void constraints::simplifyLocator(ASTNode &anchor,
     }
 
+    case ConstraintLocator::AnyPatternDecl: {
+      anchor = anchor.get<Pattern *>();
+      path = path.slice(1);
+      break;
```
I don’t think this is correct. `anchor = anchor.get<Pattern *>();` is a no-op but you’re slicing away the path element regardless. I would assume this locator can’t be simplified.

xedin replied:
Hm, the idea was to make sure that the anchor is indeed a pattern and drop the element since pattern represents the decl in this case…

ahoppen replied:
Ah, I see. I think that makes sense. So the `AnyPatternDecl` constraint locator is basically just a marker that the pattern that’s the anchor is an `AnyPattern`. In that case I’ve got a couple suggestions
- Could you add the discussion we had here in a doc comment
- Change the `anchor` reassignment to `assert(anchor.is<Pattern *>())`
- Shouldn’t we check that the anchor is an `AnyPattern`, i.e. `assert(isPattern<AnyPattern>(anchor))` (or something along those lines.

xedin replied:
Will do!

xedin replied:
Done!

ahoppen reviewed (commented):
The code completion failure might be related to the comment I put inline regarding the anchor. If not, I can take a look at it if you like.

ahoppen commented on lib/Sema/CSGen.cpp:
```diff
@@ -2293,8 +2293,10 @@ namespace {
         return setType(
             externalPatternType
                 ? externalPatternType
-                : CS.createTypeVariable(CS.getConstraintLocator(locator),
-                                        TVO_CanBindToNoEscape));
+                : CS.createTypeVariable(
+                      CS.getConstraintLocator(
+                          pattern, ConstraintLocator::AnyPatternDecl),
+                      TVO_CanBindToNoEscape | TVO_CanBindToHole));
```
Since you’re at it, I think this would read easier as

```cpp
        Type type;
        if (externalPatternType) {
          type = externalPatternType;
        } else {
          type = CS.createTypeVariable(CS.getConstraintLocator(
                                           pattern, ConstraintLocator::AnyPatternDecl),
                                       TVO_CanBindToNoEscape | TVO_CanBindToHole);
        }
        return setType(type);
```

xedin replied:
Sure, I can do that!

xedin replied:
Done!

ahoppen commented:
I just looked at the code completion failure and the problem is that in this test case we fail to resolve the type for `_` if we choose the keypath subscript overload, resulting in a fix score of 100, thus ruling out that solution. 

```swift
struct MyStruct<T> {
  static subscript(x: Int, static defValue: T) -> MyStruct<T> {
    fatalError()
  }
  subscript(x: Int, instance defValue: T) -> Int {
    fatalError()
  }
}

func test1() {
  let _ = MyStruct<Int>()[#^INSTANCE_INT_BRACKET^#
}
```

We don’t have this problem because it cannot be bound to a hole since the type variable is created by 
```cpp
varType = CS.getType(boundExpr)->getRValueType();
```

AFAICT the solution to this problem is to make variables at pattern decls as potentially incomplete, i.e. add

```cpp
    // Delay resolution of pattern decls because we prefer to determine its type
    // based on the expression it is assigned from. Also, if there is an error
    // in the expression, we want to prefer creating the direct hole inside the
    // expression, not at the pattern decl.
    if (locator->isLastElement<LocatorPathElt::PatternDecl>()) {
      return true;
    }
```

inside the `if (isHole())` of `BindingSet::isPotentiallyIncomplete`.

xedin commented:
I think it makes sense to do for `_` what is done for named in this case, if it could assume a type of initializer it should since it would regardless just indirectly…

xedin committed:
[BuilderTransform] Use correct locator for `case` subject expressions

Instead of `CTP_Initialization` transform should use `CTP_CaseStmt`
to identify the root correctly.

xedin committed:
[CSFix] Improve missing `.rawValue` fix in ambiguity conditions

xedin committed:
[CSGen] Allow `_` pattern to assume type of initializer expression

`_` pattern doesn't have a type, it's always contextual, so let's
allow it to assume a type of initializer expression instead of
creating a new type variable. This helps to makes sure that initializer
would never get a placeholder type from `_` pattern it's associated
with.

xedin commented:
@swift-ci please test

xedin committed:
[CSGen] Adjust locators for sub-patterns of parens

In cases like `.<name>(_)`, the `_` sub-pattern should not assume
locator of the parent paren, just like it doesn't in case of tuple
i.e. `.<name>(_, ...)` where each element gets a `PatternMatch(<elt>)`
locator.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility
